File: app/src/main/java/de/robinrehbein/punkt/game/GameAudio.kt
package de.robinrehbein.punkt.game

import android.content.Context
import android.media.AudioAttributes
import android.media.SoundPool
import de.robinrehbein.punkt.ui.platform.GameSounds
import java.io.File

/**
 * Chiptune-Soundeffekte für das Spiel — zur Laufzeit aus Rechteckwellen
 * synthetisiert (ChipSynth), keine Audio-Assets im Repo. Die fertigen
 * WAVs landen einmalig im Cache und laufen über einen SoundPool, damit
 * die Latenz für ein Timing-Spiel niedrig genug ist.
 *
 * Seit den Ton-Sets liegen ALLE Sets im Pool, nicht nur das gewählte:
 * Drei Sets sind zusammen keine 200 kB, und ein Wechsel muss sofort
 * hörbar sein — mit Nachladen käme die Hörprobe in der Auswahl erst,
 * wenn der Finger längst weg ist.
 */
class GameAudio(context: Context) : GameSounds {

    private val soundPool = SoundPool.Builder()
        .setMaxStreams(4)
        .setAudioAttributes(
            AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_GAME)
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .build()
        )
        .build()

    /** Ton-Set → Ereignisname → SoundPool-Id. */
    private val soundIds: Map<SoundSetId, Map<String, Int>>

    /** Stumm geschaltet? Der Screen hält das mit dem GameStore synchron. */
    override var muted: Boolean = false

    /**
     * Das gewählte Ton-Set. Der Screen hält es mit dem GameStore
     * synchron; ein noch nicht verdientes Set kommt hier gar nicht an
     * (die Auswahl lässt es nicht antippen).
     */
    override var soundSet: SoundSetId = SoundSetId.KLASSIK

    init {
        // Der Cache-Ordner ist versioniert: Ändert sich die Synthese,
        // Namen hochzählen — sonst spielen alte Dateien weiter. v2 trennt
        // die Sets in eigene Dateinamen.
        val dir = File(context.cacheDir, "sfx-v2").apply { mkdirs() }
        soundIds = SoundSetId.entries.associateWith { set ->
            ChipSynth.effects(set).mapValues { (name, samples) ->
                val file = File(dir, "${set.name}-$name.wav")
                if (!file.exists()) file.writeBytes(ChipSynth.toWav(samples))
                soundPool.load(file.path, 1)
            }
        }
    }

    override fun start() = play("start")

    /** Treffer-Blip; die Tonhöhe klettert pro 5er-Stufe eine Pentatonik hoch. */
    override fun hit(score: Int) = play("hit", rate = ChipSynth.hitRate(score))

    /** Münz-Sound; jede Serien-Stufe klingt zwei Halbtöne höher. */
    override fun perfect(streak: Int) = play("perfect", rate = ChipSynth.perfectRate(streak))

    override fun chain() = play("chain")
    override fun unlock() = play("unlock")
    override fun death() = play("death")
    override fun thud() = play("thud")
    override fun newRecord() = play("record")

    /**
     * Hörprobe für die Auswahl: die Fanfare des angetippten Sets, auch
     * wenn es gerade nicht das gewählte ist. Ein Ton-Set ohne Probe wäre
     * eine Kachel, die man kaufen soll, ohne sie gesehen zu haben — und
     * die Fanfare zeigt vom Set am meisten: Lage, Länge und Anschlag.
     */
    override fun preview(set: SoundSetId) = play("unlock", set = set)

    override fun release() = soundPool.release()

    private fun play(name: String, rate: Float = 1f, set: SoundSetId = soundSet) {
        if (muted) return
        val id = soundIds[set]?.get(name) ?: return
        soundPool.play(id, 1f, 1f, 1, 0, rate.coerceIn(0.5f, 2f))
    }
}

File: ios/Dottie/Sources/Support/GameAudio.swift
import AVFoundation
import DottieCore
import Foundation

/// Chiptune-Soundeffekte für das Spiel — zur Laufzeit aus Rechteckwellen
/// synthetisiert (ChipSynth nach der Tabelle in SoundBank), keine
/// Audio-Assets. Statt Android-SoundPool:
/// AVAudioEngine mit vorbereiteten AVAudioPCMBuffern, damit die Latenz
/// für ein Timing-Spiel niedrig genug ist. Die Tonhöhen-Varianten (Treffer-
/// Pentatonik, Perfekt-Serie) sind vorgerendert.
final class GameAudio {

    private let engine = AVAudioEngine()
    private var players: [AVAudioPlayerNode] = []
    private var nextPlayer: Int = 0

    /// Ton-Set → Klangname → Puffer. Es liegen ALLE Sets bereit: Drei
    /// Sets sind zusammen kein Speicherproblem, und die Hörprobe in der
    /// Auswahl muss sofort kommen.
    private var buffers: [SoundSetId: [String: AVAudioPCMBuffer]] = [:]

    /// Stumm geschaltet? Die Scene hält das mit dem ScoreStore synchron.
    var muted: Bool = false

    /// Das gewählte Ton-Set; die Scene hält es mit dem ScoreStore synchron.
    var soundSet: SoundSetId = .klassik

    init() {
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.ambient, options: [.mixWithOthers])
        try? session.setActive(true)

        guard let format = AVAudioFormat(
            commonFormat: .pcmFormatFloat32,
            sampleRate: Double(ChipSynth.sampleRate),
            channels: 1,
            interleaved: false
        ) else {
            return
        }

        for _ in 0..<4 {
            let player = AVAudioPlayerNode()
            engine.attach(player)
            engine.connect(player, to: engine.mainMixerNode, format: format)
            players.append(player)
        }

        for set in SoundSetId.allCases {
            var proSet: [String: AVAudioPCMBuffer] = [:]
            for event in SoundEvent.allCases where event != .hit && event != .perfect {
                proSet[event.rawValue] = GameAudio.makeBuffer(
                    ChipSynth.render(SoundBank.voice(set, event)), format: format
                )
            }
            // Treffer-Blip: fünf Pentatonik-Stufen (score % 5).
            for step in 0..<5 {
                let rate = ChipSynth.hitRate(score: step)
                proSet["hit\(step)"] = GameAudio.makeBuffer(
                    ChipSynth.render(SoundBank.voice(set, .hit), rate: rate), format: format
                )
            }
            // Münz-Sound: fünf Serien-Stufen (streak 1...5, Deckel bei 5).
            for streak in 1...5 {
                let rate = ChipSynth.perfectRate(streak: streak)
                proSet["perfect\(streak)"] = GameAudio.makeBuffer(
                    ChipSynth.render(SoundBank.voice(set, .perfect), rate: rate), format: format
                )
            }
            buffers[set] = proSet
        }

        engine.prepare()
    }

    func start() {
        play("start")
    }

    /// Treffer-Blip; die Tonhöhe klettert pro 5er-Stufe eine Pentatonik hoch.
    func hit(score: Int) {
        play("hit\(((score % 5) + 5) % 5)")
    }

    /// Münz-Sound; jede Serien-Stufe klingt zwei Halbtöne höher.
    func perfect(streak: Int) {
        play("perfect\(min(max(streak, 1), 5))")
    }

    func chain() { play("chain") }
    func unlock() { play("unlock") }
    func death() { play("death") }
    func thud() { play("thud") }
    func newRecord() { play("record") }

    /// Hörprobe für die Auswahl: die Fanfare des angetippten Sets, auch
    /// wenn es gerade nicht das gewählte ist. Ohne Probe wählt man einen
    /// Klang nach seinem Namen.
    func preview(_ set: SoundSetId) { play("unlock", set: set) }

    func release() {
        engine.stop()
    }

    private func play(_ name: String, set: SoundSetId? = nil) {
        if muted {
            return
        }
        guard let buffer = buffers[set ?? soundSet]?[name] else {
            return
        }
        if !engine.isRunning {
            do {
                try engine.start()
            } catch {
                return
            }
        }
        guard !players.isEmpty else {
            return
        }
        let player = players[nextPlayer]
        nextPlayer = (nextPlayer + 1) % players.count
        player.stop()
        player.scheduleBuffer(buffer, at: nil, options: [], completionHandler: nil)
        player.play()
    }

    private static func makeBuffer(_ samples: [Float], format: AVAudioFormat) -> AVAudioPCMBuffer? {
        let frameCount = AVAudioFrameCount(samples.count)
        guard frameCount > 0,
              let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount),
              let channel = buffer.floatChannelData else {
            return nil
        }
        buffer.frameLength = frameCount
        for i in 0..<samples.count {
            channel[0][i] = samples[i]
        }
        return buffer
    }
}
